### ISSUE TITLE
[gfx1030][Rocm5.1.1] Missing system database file: gfx1030_36.kdb

### ISSUE THREAD

**Issue by Denizzje** [ROCmSoftwarePlatform/MIOpen#1709]
Hello,

First off, I am deliberatly on ROCM 5.1.1 because the Pytorch website lists that as the current supported version. If that is the cause of this whole thing, and it can be confirmed that PyTorch works fine on ROCM 5.2.1 I'll try to reinstall it. I am just glad I had it running in the first place...

I am trying to get PyTorch to play along with my RX 6800XT. While it is somewhat functional now, I do have poor performance probably because of this database file which doesnt seem to get loaded properly?

I first tried with `sudo apt install -y miopenkernels-gfx1030-36kdb`

Which does get installed. Running it again gives: `miopenkernels-gfx1030-36kdb is already the newest version (1.1.0.50101-48).` I also rebooted.

Still getting the error, I tried the second method with `/opt/rocm-5.1.1/miopen/bin$ sudo ./install_precompiled_kernels.sh`
That does not want to run... it gives this rather odd result:

```
/opt/rocm-5.1.1/miopen/bin$ sudo ./install_precompiled_kernels.sh
using rocminfo at /usr/bin/rocminfo

sudo apt install -y miopenkernels-gfx1030-72pkg

Hit:1 http://nl.archive.ubuntu.com/ubuntu focal InRelease

Hit:2 http://security.ubuntu.com/ubuntu focal-security InRelease
               
Hit:3 http://nl.archive.ubuntu.com/ubuntu focal-updates InRelease
              
Hit:4 http://ppa.launchpad.net/cappelikan/ppa/ubuntu focal InRelease
           
Hit:5 http://nl.archive.ubuntu.com/ubuntu focal-backports InRelease
            
Hit:6 https://repo.radeon.com/amdgpu/22.10.1/ubuntu focal InRelease
            
Hit:7 https://repo.radeon.com/rocm/apt/5.1.1 ubuntu InRelease

Reading package lists... Done

Building dependency tree
       
Reading state information... Done

All packages are up to date.

Reading package lists... Done

Building dependency tree
       
Reading state information... Done

E: Unable to locate package miopenkernels-gfx1030-72pkg
```

I do not know how to intepret this as any other than its convinced I somehow have a 72 CU RX 6800? 

I do get the correct output here:
```
/opt/rocm/opencl/bin/clinfo | grep -E "gfx|units"
  Max compute units:				 36
  Name:						 gfx1030
```
Using another command I found in the git, I get the following:

```
/opt/rocm/miopen/share/miopen/db
total 861180
drwxr-xr-x 2 root root      4096 aug 18 20:15 ./
drwxr-xr-x 3 root root      4096 aug 18 20:06 ../
-rw-r--r-- 1 root root   1445888 apr  2 08:33 gfx1030_36.db
-rw-r--r-- 1 root root  36281774 apr  2 08:33 gfx1030_36.HIP.fdb.txt
-rw-r--r-- 1 root root 356933632 apr  2 08:35 gfx1030_36.kdb
-rw-r--r-- 1 root root  19560103 apr  2 08:33 gfx803_36.HIP.fdb.txt
-rw-r--r-- 1 root root  13464185 apr  2 08:33 gfx803_36.OpenCL.fdb.txt
-rw-r--r-- 1 root root  19560103 apr  2 08:33 gfx803_64.HIP.fdb.txt
-rw-r--r-- 1 root root  13464185 apr  2 08:33 gfx803_64.OpenCL.fdb.txt
-rw-r--r-- 1 root root  10739712 apr  2 08:33 gfx900_56.db
-rw-r--r-- 1 root root  23175241 apr  2 08:33 gfx900_56.HIP.fdb.txt
-rw-r--r-- 1 root root    629221 apr  2 08:33 gfx900_56.OpenCL.fdb.txt
-rw-r--r-- 1 root root  19560103 apr  2 08:33 gfx900_64.HIP.fdb.txt
-rw-r--r-- 1 root root    644832 apr  2 08:33 gfx900_64.OpenCL.fdb.txt
-rw-r--r-- 1 root root  15130624 apr  2 08:33 gfx906_60.db
-rw-r--r-- 1 root root  35790830 apr  2 08:33 gfx906_60.HIP.fdb.txt
-rw-r--r-- 1 root root   4556991 apr  2 08:33 gfx906_60.OpenCL.fdb.txt
-rw-r--r-- 1 root root  23436888 apr  2 08:33 gfx906_64.HIP.fdb.txt
-rw-r--r-- 1 root root   3322692 apr  2 08:33 gfx906_64.OpenCL.fdb.txt
-rw-r--r-- 1 root root  37810176 apr  2 08:33 gfx90878.db
-rw-r--r-- 1 root root  35027710 apr  2 08:33 gfx90878.HIP.fdb.txt
-rw-r--r-- 1 root root  28201413 apr  2 08:33 gfx90878.OpenCL.fdb.txt
-rw-r--r-- 1 root root  28110848 apr  2 08:33 gfx90a68.db
-rw-r--r-- 1 root root  49301338 apr  2 08:33 gfx90a68.HIP.fdb.txt
-rw-r--r-- 1 root root  53284864 apr  2 08:33 gfx90a6e.db
-rw-r--r-- 1 root root  52369662 apr  2 08:33 gfx90a6e.HIP.fdb.txt
```

I am on Ubuntu 20.04.4 with Linux 5.13.0-30-generic #33~20.04.1-Ubuntu SMP Mon Feb 7 14:25:10 UTC 2022 x86_64

Thanks in advance

**Comment by junliume** [comment]
@Denizzje the following should be correct:
```
sudo apt install -y miopenkernels-gfx1030-36kdb
```
The discrepancy between `36` and `72` is corrected in later releases.

Could you share more details about `poor performance`? We are interested to see the root cause of it. Thanks!

**Comment by Denizzje** [comment]
> @Denizzje the following should be correct:
> 
> ```
> sudo apt install -y miopenkernels-gfx1030-36kdb
> ```
> 
> The discrepancy between `36` and `72` is corrected in later releases.
> 
> Could you share more details about `poor performance`? We are interested to see the root cause of it. Thanks!

Greetings @junliume , 

In the mean time I actually fully uninstalled 5.1.1 (uninstall and purge, reboot), and moved to 5.2. The problem seems to persist. I did

sudo apt install -y miopenkernels-gfx1030-36kdb which actually upgraded the package.

Running it now gives: miopenkernels-gfx1030-36kdb is already the newest version (1.1.0.50200-65).

So it is installed as a package. Yet I keep having this issue. Any ideas how I can make this work?

Regarding the poor performance, I think we should have the kernel loaded in properly first before investigating it? But I can tell you the experience now. I am using YoloV5, tried both detecting and some trainings. The GPU is seen, and gets fully utilised with training (90-100% GPU and VRAM usage). Basically any action has a big delay before it, such as starting a new epoch, loading in the model itself before training, but also with just a simple inference detection. I can best describe it as it getting stuck into loading something.

Example GPU detection with YoloV5 and PyTorch Rocm:

`YOLOv5 🚀 v6.2-13-gc0e7a77 Python-3.8.10 torch-1.12.1+rocm5.1.1 CUDA:0 (AMD Radeon RX 6800 XT, 16368MiB)

Fusing layers... 
MIOpen(HIP): Warning [SQLiteBase] Missing system database file: gfx1030_36.kdb Performance may degrade. Please follow instructions to install: https://github.com/ROCmSoftwarePlatform/MIOpen#installing-miopen-kernels-package
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
image 1/2 /home/denizz/yolov5/data/images/bus.jpg: 640x480 4 persons, 1 bus, Done. (0.188s)
image 2/2 /home/denizz/yolov5/data/images/zidane.jpg: 384x640 2 persons, 2 ties, Done. (0.196s)
Speed: 0.3ms pre-process, 192.0ms inference, 41.0ms NMS per image at shape (1, 3, 640, 640)
`
Vs running it on CPU (Ryzen R7 5800X)

`YOLOv5 🚀 v6.2-13-gc0e7a77 Python-3.8.10 torch-1.12.1+rocm5.1.1 CPU

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
image 1/2 /home/denizz/yolov5/data/images/bus.jpg: 640x480 4 persons, 1 bus, Done. (0.047s)
image 2/2 /home/denizz/yolov5/data/images/zidane.jpg: 384x640 2 persons, 2 ties, Done. (0.035s)
Speed: 2.9ms pre-process, 40.9ms inference, 2.3ms NMS per image at shape (1, 3, 640, 640)
Results saved to runs/detect/exp12
`
Similar delays also happen with just the PyTorch MNIST example. It takes a while before it even starts the training.

**Comment by junliume** [comment]
@Denizzje could you please check the following:

move `gfx1030_36.kdb` from `/opt/rocm/miopen/share/miopen/db` to `/opt/rocm/share/miopen/db`

This is caused by some folder re-organization changes in ROCm for 5.2 and should have been fixed in the latest `develop` branch.

**Comment by Denizzje** [comment]
There was no file there when I put in the move command, 

and when taking a look they folders and files were already there post reorganisation (I did do a full reinstall when going from 5.1.1 to 5.2.0).

![image](https://user-images.githubusercontent.com/25279882/185502568-70476d1e-0613-4cc5-832c-4ec9076ad2f2.png)

But... I ""Solved"" the problem. I built PyTorch from source. And by chance, I found this issue as well:  https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1391

I believe this is the cause of what I experienced. Currently, it seems that PyTorch + Rocm from Pip wheel is not usuable because of this. As having to build from source is a big inconvieniance, I hope this can be looked at.

I can now properly evaluate using ROCM + PyTorch for now however, perhaps it is better to close this exact issue.

**Comment by junliume** [comment]
Thanks! @Denizzje I have pinned the issue with PyTorch Wheel and KDB to higher priority and will assign developers to work on it.

Please report any issues in your use cases if any, and any suggestions or recommendations are highly appreciated!

**Comment by atamazov** [comment]
IIRC adding missing kdb files to PyTorch wheels is possible only via rebuilding the wheels, because kdb is embedded during the build step.

/cc @JehandadKhan

**Comment by Denizzje** [comment]
Hi @junliume  ,  I see there has been a ROCM 5.3 with Ubuntu 22.04 support. Which is good because I feel like taking a time capsule when having to boot 20.04 ;). 

 Just a quick follow up question if there has been some progress with this issue or do I need to prepare to compile PyTorch from source when reinstalling Ubuntu and ROCM 5.3? Id rather set it up correctly at once so if there has been no fix for the Pytorch wheels to find the kdb correctly then atleast I know that I should compile myself. And ofcourse happy to test if there is.

**Comment by atamazov** [comment]
@Denizzje Most likely, your question (together with the issue of missing embedded pre-compiled kernel db for gfx1030/36 in PyTorch) should be addressed to Frameworks team.

@junliume @sunway513 @JehandadKhan What do you think?

**Comment by cstueckrath** [comment]
any updates on this?

**Comment by atamazov** [comment]
@cstueckrath AFAICS this issue should be resolved at the PyTorch wheel level, see https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1709#issuecomment-1239901804 and https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1709#issuecomment-1269069242. I recommend closing this ticket.

/cc @junliume

**Comment by jmcelroy01** [comment]
I have a similar issue when using the Automatic1111 Stable Diffusion. On launch, it says `Warning [SQLiteBase] Missing system database file: gfx1030_14.kdb Performance may degrade.`, but there is no such file in `/opt/rocm-5.4.0/share/miopen/db/`. I am using `HSA_OVERRIDE_GFX_VERSION=10.3.0` to make Pytorch work with the 6700S GPU.

**Comment by Dajinu** [comment]
> There was no file there when I put in the move command for gfx1030_36.kdb.
> 
> and when taking a look the mentioned folder to move the file in, were already there in their post reorganisation format (I did do a full reinstall when going from 5.1.1 to 5.2.0).
> 
> ![image](https://user-images.githubusercontent.com/25279882/185502568-70476d1e-0613-4cc5-832c-4ec9076ad2f2.png)
> 
> But... I ""Solved"" the problem. I built PyTorch from source. And by chance, I found this issue as well: #1391
> 
> I believe this is the cause of what I experienced. Currently, it seems that PyTorch + Rocm from Pip wheel is not usuable because of this. As having to build from source is a big inconvieniance, I hope this can be looked at.
> 
> The kernell warning message is gone, and everything responds much more like I would expect.
> 
> I can now properly evaluate using ROCM + PyTorch for now however, perhaps it is better to close this exact issue.

Hello @Denizzje ! I am trying to build pytorch with included miopen kernel cache file with no success. I'm so desperate that I asked ChatGPT how to achieve this. He gave me this line: `python setup.py build -DUSE_MIOPEN=ON -DMIOPEN_USER_DB_PATH=/opt/rocm/share/miopen/db`
But that returns an error: option -D not recognized
I have no coding background and I'm trying so hard to understand all this but it's really difficult for me.
Please, can you tell me how you did to build pytorch with the kernel cache file included? I'd really be grateful! Thank you in advance!

**Comment by atamazov** [comment]
@Dajinu 
> I am trying to build pytorch with included miopen kernel cache file with no success. Please, can you tell me how you did to build pytorch with the kernel cache file included? I'd really be grateful! Thank you in advance!

This is non trivial. But you can safely ignore this warning. This only affects initial runs of neural networks on your system, because MIOpen (1) builds kernels when necessary and (2) supports per-user binary cache for all GPUs.

If you still wish to cope with this, then please ask your question in https://github.com/pytorch/pytorch.

@junliume I recommend closing this due to https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1709#issuecomment-1339876069

**Comment by Denizzje** [comment]
I will close this myself @atamazov  , I have been unable to work on getting a ROCM + PyTorch setup going and might take a while before I do again due to time constraints. We are already at ROCM 5.4.3 now so no need to have a ROCM 5.1.1 issue lingering on.

Before I close, could you check if I understand your recommendations correctly: The missing kernel files in the pip wheel of PyTorch is not something which can be adressed by the AMD/ROCM team and to see if we can get it solved I will need to raise the issue with PyTorch itself?

@Dajinu strongly recommend to NOT use chatgpt for questions like that, as it is likely using outdated information due to it not knowing anything past a point in 2021, while frameworks like ROCM and PyTorch evolve all the time. I do not have a coding background myself and had never built an application before. I made various mistakes when building PyTorch back then but it came down to me not reading the instructions properly or skipping steps. I unfortunally cannot debug your attempt, it is too long ago by now. Not meant as critique just advice from a fellow novice ;).

**Comment by atamazov** [comment]
@Denizzje 
> The missing kernel files in the pip wheel of PyTorch is not something which can be adressed by the AMD/ROCM team...

Not exactly. IIRC the AMD/ROCm guys participate in PyTorch development.

> ...and to see if we can get it solved I will need to raise the issue with PyTorch itself?

Like that. We (MIOpen team) already provide `gfx1030_36.kdb` and also make sure that this file is installed on the system during normal installation. But it is up to applications that use database embedding to ensure embedding of proper databases. I am not fully aware of who is responsible for building pip wheel of PyTorch, but these are whom this issue should be addressed to.

AFAICS there is a ticket already, https://github.com/pytorch/pytorch/issues/94482.

**Comment by Denizzje** [comment]
Thanks for clarifying @atamazov . As a novice in this field it can all be overwhelming :-) but I think I get it now. I will now close the issue so this doesnt keep popping up randomly over the course of months. I will direct future inquiries about this to PyTorch.
